### ISSUE TITLE
observations CRUD

### ISSUE THREAD

**Issue by mjy** [SpeciesFileGroup/matrix_row_coder#1]
```
  when I update an observation
       then I should the corresponding request to /observations or /observations:id
```

**Comment by BrianWalters** [comment]
Hey, @mjy, what exactly does the PUT request look like?

Let's say for a qualitative descriptor that has a character state id of 84. Would that be:

/observations.json

with a JSON payload of:

```
{
  "otu_id": 1,
  "descriptor_id": 2,
  "project_id": 1,
  "character_state_id": 84
}
```

Am I thinking about this right?

**Comment by mjy** [comment]
Exactly right, as a POST.

**Comment by mjy** [comment]
BDub-

If PATCH/PUT/DESTROY you pass `id: 1` as well, somewhat obviously.

**Comment by BrianWalters** [comment]
For clarification:

POST to /observations.json creates a new observation
DELETE to /observations/1.json removes the observation with id 1.
PUT to observations/1.json changes the observation with id 1.

Is that right?

**Comment by mjy** [comment]
Yes, though there might be some trickery to using patch, not put, will
check in a bit.

M

**Comment by BrianWalters** [comment]
I'm also thinking that Sample and Continuous observations should include a delete button next to their values. Emptying the inputs feels ambiguous.

**Comment by mjy** [comment]
Yes!

**Comment by BrianWalters** [comment]
My client side code is almost done. My strategy now is this:

Qualitative and Presence observations don't have an "update" action with them. They either exist or they don't. So they only handle create, read, and delete. Delete and create will occur after their checked state is updated.

Sample and Continuous observations have full CRUD. Updates can happen when one of their properties is updated in the state. Remove button will delete them and null out their properties in the state. Updating properties on an empty observation that doesn't have an id will trigger a create when the observation is saved.

**Comment by BrianWalters** [comment]
Creating Qualitative and Presence looks good.

What is required for the create payload for Sample and Continuous?

**Comment by mjy** [comment]
Try these in postman (same endpoint):

Sample:

Only serialize those "sample" attributes that have data:

{
    "otu_id": "1",
    "project_id": "1",
    "token": "DPpUOSv_VHQjqsy1o2itYg",
    "descriptor_id": "123123",
    "type": "Observation::Sample",
    "sample_n": "2"
   ....  and these ...
    "sample_min": ,
    "sample_max":
    "sample_median":
    "sample_mean":
    "sample_units": ,
    "sample":
    "sample_standard_error":
}


PresenceAbsence:

{
    "otu_id": "1",
    "project_id": "1",
    "token": "DPpUOSv_VHQjqsy1o2itYg",
    "descriptor_id": "99",
   "presence": "1"
    "type": "Observation::PresenceAbsence"
}

If "1" (and "0") don't work try "true" and "false".  This reminds me that
the form really has to be three state.  (present, absent, and something
like "not provided").


Make sense?

**Comment by BrianWalters** [comment]
Yeah that makes sense. I have it rigged up now that a presence observation either exists or it doesn't. So unchecking a presence descriptor's checkbox will delete the presence observation. If a presence observation gets switched to presence absence, does that mean creating a new observation or editing that same presence observation?

Another thing with samples: the mean, median, and standard error sound like calculations. Should those be done client side and sent up?

**Comment by mjy** [comment]
On Wed, May 3, 2017 at 11:46 AM, Brian Walters <notifications@github.com>
wrote:

> Yeah that makes sense. I have it rigged up now that a presence observation
> either exists or it doesn't. So unchecking a presence descriptor's checkbox
> will delete the presence observation. If a presence observation gets
> switched to presence absence, does that mean creating a new observation or
> editing that same presence observation?
>
Edit the same observation.


> Another thing with samples: the mean, median, and standard error sound
> like calculations. Should those be done client side and sent up?
>

They are not calculations, they are all text values.

M

**Comment by BrianWalters** [comment]
Hey Matt, I'm getting the same CORS error when posting observations

![screen shot 2017-05-03 at 15 48 35](https://cloud.githubusercontent.com/assets/348655/25680993/4e4dcf5e-3018-11e7-8c0f-8e4f239f2c8a.png)

Works in Postman though

**Comment by mjy** [comment]
Are you hitting it with .json?  Why does it say index.js?

M

**Comment by mjy** [comment]
BDub-

Make sure your request this in the header:
Content-Type →application/json; charset=utf-8
